### PR TITLE
Have the wallet give out destinations instead of keys

### PR DESCRIPTION
**Pull request by achow101**
The wallet should give out destinations instead of keys. It should be the one that handles the conversion from key to destination and the setting of the label, not the caller. In order to do this, two new member functions are introduced `GetNewDestination()` and `GetNewChangeDestination()`. Additionally, `CReserveKey` is changed to be `ReserveDestination` and represents destinations whose keys can be returned to the keypool.

### CODE REVIEW

File: src/wallet/wallet.h [hunk 1]
```diff
@@ -1123,6 +1123,8 @@ class CWallet final : public CCryptoKeyStore, private interfaces::Chain::Notific
 
     std::set<CTxDestination> GetLabelAddresses(const std::string& label) const;
 
+    bool GetNewAddress(const OutputType type, const std::string label, CTxDestination& dest, std::string& error);
```
Comment (promag) [inline src/wallet/wallet.h]: 2010b1a13e6b7eaa681cc62ae58079ccbec72816

This should be atomic or add `EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)` annotation?

Also, could make `GetKeyFromPool` private?

**Review by promag — COMMENTED** [review]
Concept ACK

You have a typo in commit 7d014e492cdcca847cc0e46f0457b1bc89831f90 message.

**Review by ryanofsky — COMMENTED** [review]
Concept ACK. Will review more closely, but this seems like a nice, straightforward change.

File: src/wallet/wallet.h (hunk 1 above)
Comment (achow101) [reply src/wallet/wallet.h]: Done. Also fixed typo

File: src/interfaces/wallet.cpp [hunk 2]
```diff
@@ -140,9 +140,10 @@ class WalletImpl : public Wallet
     void abortRescan() override { m_wallet->AbortRescan(); }
     bool backupWallet(const std::string& filename) override { return m_wallet->BackupWallet(filename); }
     std::string getWalletName() override { return m_wallet->GetName(); }
-    bool getKeyFromPool(bool internal, CPubKey& pub_key) override
+    bool getNewAddress(const OutputType type, const std::string label, CTxDestination& dest) override
     {
-        return m_wallet->GetKeyFromPool(pub_key, internal);
+        std::string error;
+        return m_wallet->GetNewAddress(type, label, dest, error);
```
Comment (MarcoFalke) [inline src/interfaces/wallet.cpp]:
```
interfaces/wallet.cpp:146:16: error: calling function 'GetNewAddress' requires holding mutex 'm_wallet->cs_wallet' exclusively [-Werror,-Wthread-safety-analysis]

Comment (achow101) [reply src/interfaces/wallet.cpp]: Fixed

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -2791,19 +2791,16 @@ bool CWallet::CreateTransaction(interfaces::Chain::Lock& locked_chain, const std
...
                 }
-                CPubKey vchPubKey;
-                bool ret;
-                ret = reservekey.GetReservedKey(vchPubKey, true);
+                CTxDestination dest;
+                const OutputType change_type = TransactionChangeType(coin_control.m_change_type ? *coin_control.m_change_type : m_default_change_type, vecSend);
+                bool ret = reserveaddr.GetReservedAddress(change_type, dest, true);
                 if (!ret)
                 {
                     strFailReason = _("Keypool ran out, please call keypoolrefill first");
```
Comment (laanwj) [inline src/wallet/wallet.cpp]: nit: the message is translated here, but not in the new occurences. As this looks like a message that will be returned over the RPC interface (referring to a RPC command, at least). Maybe better to not translate it here either?

**Comment by laanwj** [comment]
Concept ACK
(my remark was going to be that the wallet should return Destinations, not addresses, but that's exactly that this does)

**Comment by Sjors** [comment]
Concept ACK.

In line with @laanwj's comment, maybe rename `GetNew(Change)Address` to `GetNew(Change)Destination`, and `ReserveAddress` to `ReserveDestination`.

`GetReservedAddress` is still used in two places; maybe kill it in this PR?

Travis seems unhappy about a locking issue. On macOS I also get some warnings:
Schermafbeelding 2019-07-09 om 16 19 25

**Comment by achow101** [comment]
> In line with @laanwj's comment, maybe rename `GetNew(Change)Address` to `GetNew(Change)Destination`, and `ReserveAddress` to `ReserveDestination`.

I've renamed it.

> `GetReservedAddress` is still used in two places; maybe kill it in this PR?

It is?

> Travis seems unhappy about a locking issue. On macOS I also get some warnings:

Should be fixed now.

File: src/wallet/wallet.cpp (hunk 3 above)
Comment (achow101) [reply src/wallet/wallet.cpp]: I've made it untranslated.

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -3513,6 +3513,26 @@ bool CWallet::GetKeyFromPool(CPubKey& result, bool internal)
     return true;
 }
 
+bool CWallet::GetNewDestination(const OutputType type, const std::string label, CTxDestination& dest, std::string& error)
+{
```
Comment (instagibbs) [inline src/wallet/wallet.cpp]: micro-nit: can clear `error` here.

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -3728,25 +3725,29 @@ bool CReserveKey::GetReservedKey(CPubKey& pubkey, bool internal)
...
     assert(vchPubKey.IsValid());
-    pubkey = vchPubKey;
+    pwallet->LearnRelatedScripts(vchPubKey, type);
+    address = GetDestinationForKey(vchPubKey, type);
+    dest = address;
     return true;
 }
 
-void CReserveKey::KeepKey()
+void ReserveDestination::KeepAddress()
```
Comment (instagibbs) [inline src/wallet/wallet.cpp]: We're using "destination" in `GetReservedDestination` but "address" here. We should pick one?

File: src/wallet/wallet.cpp [hunk 6]
```diff
@@ -3728,25 +3725,29 @@ bool CReserveKey::GetReservedKey(CPubKey& pubkey, bool internal)
...
 {
     if (nIndex != -1)
         pwallet->KeepKey(nIndex);
     nIndex = -1;
     vchPubKey = CPubKey();
+    address = CNoDestination();
 }
 
-void CReserveKey::ReturnKey()
+void ReserveDestination::ReturnAddress()
```
Comment (instagibbs) [inline src/wallet/wallet.cpp]: We're using "destination" in `GetReservedDestination` but "address" here. We should pick one?

File: src/wallet/wallet.cpp [hunk 7]
```diff
@@ -3530,6 +3530,22 @@ bool CWallet::GetNewDestination(const OutputType type, const std::string label,
     return true;
 }
 
+bool CWallet::GetNewChangeDestination(const OutputType type, CTxDestination& dest, std::string& error)
+{
```
Comment (instagibbs) [inline src/wallet/wallet.cpp]: nit: same comment about clearing the error string

**Review by instagibbs — COMMENTED** [review]
concept ACK

File: src/wallet/wallet.cpp (hunk 4 above)
Comment (achow101) [reply src/wallet/wallet.cpp]: Does it matter?

File: src/wallet/wallet.cpp (hunk 5 above)
Comment (achow101) [reply src/wallet/wallet.cpp]: Whoops, fixed

File: src/wallet/wallet.cpp (hunk 6 above)
Comment (achow101) [reply src/wallet/wallet.cpp]: Fixed

File: src/wallet/wallet.cpp (hunk 4 above)
Comment (instagibbs) [reply src/wallet/wallet.cpp]: micro-matters in case someone calls the function twice and doesn't clear the error themself? :)

**Comment by instagibbs** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/16237/commits/c84cbd50ce9d0418c6f71d56995e03a3ed7f56bc

File: src/wallet/wallet.cpp (hunk 4 above)
Comment (achow101) [reply src/wallet/wallet.cpp]: Done

File: src/wallet/wallet.cpp (hunk 7 above)
Comment (achow101) [reply src/wallet/wallet.cpp]: Done

Comment (achow101) [reply src/wallet/wallet.cpp]: Done

**Comment by instagibbs** [comment]
re-utACK https://github.com/bitcoin/bitcoin/pull/16237/commits/8e7f930828a9f8f9be1c90ff45e3fdfef1980eaf

only changes are clearing error messages at beginning of `GetNew*`

**Comment by sipa** [comment]
ACK 8e7f930828a9f8f9be1c90ff45e3fdfef1980eaf. Concept ACK as this gives a much cleaner abstraction to work with, and light code review ACK.

It's bit confusing that constructing a (filled) ReserveDestination does not actually require a destination type, preventing the interface from being able to select where to draw change from based on the address type desired. Perhaps something for a follow-up?

**Comment by laanwj** [comment]
ACK 8e7f930828a9f8f9be1c90ff45e3fdfef1980eaf

**Comment by Sjors** [comment]
The macOS warnings I saw before are indeed gone now.

**Pull request merged**